service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java

```diff
@@ -0,0 +1,110 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+@Service
+public class DraftCMOService {
+
+    private final ObjectMapper mapper;
+    private final DateFormatterService dateFormatterService;
+
+    @Autowired
+    public DraftCMOService(DateFormatterService dateFormatterService, ObjectMapper mapper) {
+        this.mapper = mapper;
+        this.dateFormatterService = dateFormatterService;
+    }
+
+    public DynamicList getHearingDatesDynamic(CaseDetails caseDetails) {
+        Map<String, Object> caseDataMap = caseDetails.getData();
+        CaseData caseData = mapper.convertValue(caseDataMap, CaseData.class);
+
+        List<Element<HearingBooking>> hearingDetails = caseData.getHearingDetails();
+
+        DynamicList hearingDatesDynamic = buildDynamicListFromHearingDetails(hearingDetails);
+        CaseManagementOrder caseManagementOrder = caseData.getCaseManagementOrder();
+
+        prePopulateHearingDateSelection(hearingDetails, hearingDatesDynamic, caseManagementOrder);
+
+        return hearingDatesDynamic;
+    }
+
+    private void prePopulateHearingDateSelection(
+        List<Element<HearingBooking>> hearingDetails,
+        DynamicList hearingDatesDynamic,
+        CaseManagementOrder caseManagementOrder) {
+        if (caseManagementOrder != null) {
+            String hearingDateId = ObjectUtils.isEmpty(caseManagementOrder) ? ""
+                : caseManagementOrder.getHearingDateId();
+            if (!isEmpty(hearingDateId)) {
+                // There was a previous hearing date therefore we need to remap it
+                String date = hearingDetails.stream()
+                    .filter(Objects::nonNull)
+                    .filter(element -> element.getId().toString().equals(hearingDateId))
+                    .findFirst()
+                    .map(element -> convertDate(element.getValue().getDate()))
+                    .orElse("");
+
+                DynamicListElement listElement = DynamicListElement.builder()
+                    .label(date)
+                    .code(hearingDateId)
+                    .build();
+
+                hearingDatesDynamic.setValue(listElement);
+                if (!hearingDatesDynamic.getListItems().contains(listElement)) {
+                    hearingDatesDynamic.getListItems().add(listElement);
+                }
+            }
+        }
+    }
+
+    public DynamicList buildDynamicListFromHearingDetails(List<Element<HearingBooking>> hearingDetails) {
+        List<HearingDate> hearingDates = hearingDetails
+            .stream()
+            .map(element -> new HearingDate(element.getId(), element.getValue().getDate()))
+            .collect(Collectors.toList());
+
+        return DynamicList.toDynamicList(hearingDates, DynamicListElement.EMPTY);
+    }
+
+    public String convertDate(LocalDate date) {
```

Comment:
Should this not be private?

Reply:
☝️

Reply:
does this method not exist in dateFormatterService?

Reply:
@amitk2281 this doesn't seem to have been resolved

Reply:
This method is using dateFormatterService. This method is just to make it clean. In place of writing `dateFormatterService.formatLocalDateToString(date, FormatStyle.MEDIUM)` line more than one places in code, just moved to one method.

Reply:
Its not private as reusing the same method in DraftCMOServiceTest else need to use the write same code at various places in tests.

Reply:
I don't think this is correct. It should be a private method.

Reply:
you should never write production code purely to use in tests

Reply:
removed method and moved to DateFormatterService